### ISSUE TITLE
"decoding error" from nativemethods.m

### ISSUE THREAD

**Issue by ddyer0** [codenameone/CodenameOne#2217]
This is a new problem, first noticed a week ago, so probably corresponds to some recent
change in the UTF8 decoder.   Reading a plain asciii file over https, this error is thrown.
The only place this exception occurs is in nativemethods.m

> [online.game.Game] 0:0:0,0 - Exception: java.lang.RuntimeException - Decoding Error
java.lang.RuntimeException
    at online_game_sgf_sgf_reader.sgf_parse:62
    at online_game_sgf_sgf_reader.sgf_parse:302
    at online_game_sgf_sgf_reader.parse_sgf_file:212
    at online_game_commonCanvas.doLoadUrl:783
    at online_game_Game.setServerFile:95
    at online_game_Game.deferredSet:579
    at online_common_commonPanel.runStep:579
    at online_game_Game.runStep:268
    at online_game_Game.run:51
    at java_lang_Thread.runImpl:153

The same file, read from the same build, under other circumstances reads fine, but once
wedged the reader seems to be wedged for good.  I haven't managed to reproduce the
problem, but it occurs occasionally.  Other files read through this mechanism might
contain some UTF8 encodings, but the file in question does not.

**Comment by shannah** [comment]
Can you post the file?

**Comment by ddyer0** [comment]
The file is not relevant, it's just ascii.  But here's one that has been implicated.

**Comment by shannah** [comment]
Then I can't reproduce the issue.  I have many ascii files in my test suite.  Non-reproducible issues take much longer to fix.

**Comment by ddyer0** [comment]
My working hypothesis is that there's some stale state being carried over from a previous file.

**Comment by shannah** [comment]
What is on line 6 of sgf_parse()?

**Comment by ddyer0** [comment]
that's not a line number.  It would be a call to read.

**Comment by shannah** [comment]
You mean you don't have the source?  That should be a line number

**Comment by ddyer0** [comment]
..and the file would be a URL, for which the stream is the data part, ultimately associated with ConnectionRequest.getResponseData().  Getting http data is quite convoluted. I believe this is
the code that produces the stream that is throwing the error.
`	//
	// this is a callback when the data is available.
	//
	public void actionPerformed(NetworkEvent evt) 
	{
        byte[] c_result = getResponseData();
        ByteArrayInputStream in = c_result==null ? null : new ByteArrayInputStream(c_result);
        responseData = in;
	}
`

**Comment by shannah** [comment]
The exception is thrown inside a java.lang.String() constructor that takes a byte array as a parameter.  I don't see that on that line.  Perhaps that occurs on the next line.

In any case, the UTF8-decoder is declaring that the bytes that are passed to the string contain an invalid byte sequence.  Note that ASCII != ISOLatin1.  If the string contains latin-1 characters, you'd need to specify that in the charset parameter ("ISO-8859-1"), or it will default to UTF-8 - and it may blow up if it reaches a character that isn't UTF-8.

If you're 100% sure it's ASCII, you can use "US-ASCII" as the charset, explicitly, and that will by-pass the UTF-8 decoder.

**Comment by ddyer0** [comment]
As I said earlier in this thread, some files will include UTF-8 encodings.  This file doesn't.  My assumption is that the UTF8 decoder is not completely initialized, or has been damaged in some previous accident.  Since this is raw c code, there's lots of opportunity for things like that.

**Comment by shannah** [comment]
That's a possibility.  But I find it more likely that the input isn't valid UTF-8.  If it is a different encoding, you should use the charset parameter to declare that.

**Comment by ddyer0** [comment]
My previous attachment didn't work for some reason (probably operator error) so here's 
one of the culprit files again.  As you can see, it's just ascii, and under other circumstances
the same file reads just fine in its original context, which is 
https://boardspace.net/rithmomachy/rithmomachygames/RM-Dumbot-stepanzo-2015-10-29-2124.sgf
[RM-Dumbot-stepanzo-2015-10-29-2124.txt](https://github.com/codenameone/CodenameOne/files/1300299/RM-Dumbot-stepanzo-2015-10-29-2124.txt)

**Comment by shannah** [comment]
Yes.  File looks fine.  I haven't tried it on my end yet, but you say it doesn't fail every time?

**Comment by ddyer0** [comment]
Correct.  I've been looking, but haven't found any trigger for the broken state.

**Comment by ddyer0** [comment]
this may be related to #3919 which was recently fixed.
